tests/test_custom_attributes.py

```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
+import pytest
+from mock import Mock
+
+from ansible.module_utils.basic import AnsibleModule
+
+from miqclient.api import API
+import manageiq_custom_attributes
+
+
+MANAGEIQ_HOSTNAME = "http://miq.example.com"
+PROVIDER_NAME = "openshift01"
+PROVIDER_HOSTNAME = "os01.example.com"
+PROVIDER_ID = 266
+EXISTING_CA = {"id": 6226, "name": "ca1", "value": "value 1"}
+NEW_CA = {"id": 6227, "name": "ca2", "value": "value 2"}
+UPDATED_CA_VALUE = "new value"
+
+
+GET_RETURN_VALUES = {
+    'no_cas': {},
+    'ca_exist': {
+        "custom_attributes": [{
+            "href": "{miq_hostname}/api/providers/{provider_id}/custom_attributes/{ca_id}".format(miq_hostname=MANAGEIQ_HOSTNAME, provider_id=PROVIDER_ID, ca_id=EXISTING_CA['id']),
+            "id": EXISTING_CA['id'],
+            "section": "metadata",
+            "name": EXISTING_CA['name'],
+            "value": EXISTING_CA['value'],
+            "resource_type": "ExtManagementSystem",
+            "resource_id": PROVIDER_ID,
+            "source": "EVM",
+            "serialized_value": EXISTING_CA['value']
+        }]
+    }
+}
+POST_RETURN_VALUES = {
+    'added_ca': {
+        'results': [{
+            "resource_id": PROVIDER_ID,
+            "section": "metadata",
+            "name": NEW_CA['name'],
+            "id": NEW_CA['id'],
+            "serialized_value": NEW_CA['value'],
+            "value": NEW_CA['value'],
+            "resource_type": "ExtManagementSystem",
+            "source": "EVM"
+        }]
+    },
+    'updated_ca': {
+        'results': [{
+            "resource_id": PROVIDER_ID,
+            "section": "metadata",
+            "name": EXISTING_CA['name'],
+            "id": EXISTING_CA['id'],
+            "serialized_value": UPDATED_CA_VALUE,
+            "value": UPDATED_CA_VALUE,
+            "resource_type": "ExtManagementSystem",
+            "source": "EVM"
+        }]
+    }
+}
+
+
+@pytest.fixture(autouse=True)
+def miq_api_class(monkeypatch):
+    miq_api_class = Mock(spec=API)
+    monkeypatch.setattr("manageiq_custom_attributes.MiqApi", miq_api_class)
+    yield miq_api_class
+
+
+@pytest.fixture
+def the_provider():
+    the_provider = Mock()
+    the_provider.name = PROVIDER_NAME
+    yield the_provider
+
+
+@pytest.fixture
+def miq_ansible_module():
+    miq_ansible_module = Mock(spec=AnsibleModule)
+    yield miq_ansible_module
+
+
+class AnsibleModuleFailed(Exception):
+    pass
+
+
+@pytest.fixture()
+def miq(miq_api_class, miq_ansible_module, the_provider):
+    # miq_ansible_module.params = {'metrics': True}
+
+    def fail(msg):
```

Comment:
It will be nice to have failure test, bad arguments / update failure / deletion / CA already exists